tools/blkalgn.py

```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/env python
+# SPDX-License-Identifier: Apache-2.0
+#
+# Block alignment observability tool.
+#
+# Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved.
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 06-Nov-2023   Daniel Gomez   Created this.
+from __future__ import (
+    absolute_import, division, unicode_literals, print_function
+)
+from bcc import BPF
+import argparse
+import time
+
+examples = """examples:
+  blkalgn                             # Observe all blk commands
+  blkalgn --disk nvme9n1              # Observe all commands on 9th NVMe node
+  blkalgn --ops read                  # Observe read commands on all NVMe
+  blkalgn --ops write                 # Observe write commands on all NVMe
+  blkalgn --ops write --disk nvme9n1  # Observe write commands on 9th NVMe node
+  blkalgn --debug                     # Print eBPF program before observe
+  blkalgn --trace                     # Print NVMe captured events
+  blkalgn --interval 0.1              # Poll data ring buffer every 100 ms
+"""
+
+parser = argparse.ArgumentParser(
+    description="Block commands observer tool",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples,
+)
+parser.add_argument(
+    "-d",
+    "--disk",
+    type=str,
+    help="capture commands for this block device node only"
+)
+parser.add_argument(
+    "-o",
+    "--ops",
+    type=str,
+    help="capture this command operation only"
+)
+parser.add_argument("--debug", action="store_true", help="debug")
+parser.add_argument(
+    "--trace",
+    action="store_true",
+    help="trace block captured commands"
+)
+parser.add_argument(
+    "--interval",
+    type=float,
+    help="polling interval"
+)
+
+args = parser.parse_args()
+
+# define BPF program
+bpf_text = """
+#include <uapi/linux/ptrace.h>
+#include <linux/blk-mq.h>
+
+struct data_t {
+    u32 pid;
+    char comm[TASK_COMM_LEN];
+    char disk[DISK_NAME_LEN];
+    u32 op;
+    u32 len;
+    u32 lba;
+    u32 algn;
+};
+
+BPF_HISTOGRAM(block_len, u32, 64);
+BPF_HISTOGRAM(algn, u32, 64);
+BPF_ARRAY(counts, u64, 1);
+BPF_RINGBUF_OUTPUT(events, 8);
+
+/* local strcmp function, max length 16 to protect instruction loops */
+#define CMPMAX	16
+
+static int local_strcmp(const char *cs, const char *ct)
+{
+    int len = 0;
+    unsigned char c1, c2;
+
+    while (len++ < CMPMAX) {
+        c1 = *cs++;
+        c2 = *ct++;
+        if (c1 != c2)
+            return c1 < c2 ? -1 : 1;
+        if (!c1)
+            break;
+    }
+    return 0;
+}
+"""
+
+bpf_text_disk_filter = ""
+if args.disk:
+    bpf_text_disk_filter = """
+        if (local_strcmp(req->q->disk->disk_name, "{disk}"))
+            return;
+    """.format(
+        disk=args.disk
+    )
+
+bpf_text_ops_filter = ""
+# Operation dictionary. Full list of operations at Linux kernel
+# 'include/linux/blk_types.h' header file.
+blk_ops = {
+    0: "read",
+    1: "write",
+    "read": 0,
+    "write": 1,
+}
```

Comment:
Add more `enum req_op` op codes? I got this error:
```
vda        write    4096     8668088    5774       kworker/u8:3     4096    
Exception ignored on calling ctypes callback function: <function RingBuf.open_ring_buffer.<locals>.ringbuf_cb_ at 0xffff70c50f40>
Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/bcc/table.py", line 1260, in ringbuf_cb_
    ret = callback(ctx, data, size)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/root/test/nvmecmd/./blkalgn.py", line 186, in capture_event
    print_event(event)
  File "/root/test/nvmecmd/./blkalgn.py", line 194, in print_event
    blk_ops[event.op],
    ~~~~~~~^^^^^^^^^^
KeyError: 34
```
Add all the ops as for general purpose block device:
```
DISK       OPS      LEN      LBA        PID        COMM             ALGN    
sr0        DRV_IN   8        4294967295 6322       kworker/3:0      4096    
vda        Write    4096     8668056    6684       kworker/u8:3     4096    
^C
```
Most interested op should be write for alignment checking, other sector related ops are read, discard (unmap/trim), write zeros and probably ZNS zones. Maybe just check sector size != 0 and filter those unrelated ops.

Reply:
@xutao323, I've added all the requests possible and convert them to Pascal Case as the `biolatency` tool. That makes the `--ops write` be `--ops Write`. But input does not match, then `blkalgn` prints a helper list:

```
blkalgn.py --trace --ops write

Operation does not exist. Please, introduce any valid operation
Read
Write
Flush
Discard
SecureErase
WriteZeroes
ZoneOpen
ZoneClose
ZoneFinish
ZoneAppend
ZoneReset
ZoneResetAll
DrvIn
DrvOut
Last
```

Reply:
I'm closing this. Please, reopen the conversation if you think the solution is not appropriate. Thanks!